Userland/Libraries/LibJS/Parser.cpp

```diff
@@ -241,12 +267,98 @@ class ScopePusher {
             }
         }
 
+        for (auto& it : m_identifier_groups) {
+            auto const& identifier_group_name = it.key;
+            auto& identifier_group = it.value;
+
+            if (identifier_group_name == "arguments"sv) {
+                // NOTE: arguments is a special variable that should not be treated as a candidate to become local
+                continue;
+            }
+
+            bool scope_has_declaration = false;
+            MUST(m_node->for_each_var_declared_name([&](auto const& name) {
+                if (m_function_parameters.has_value() && m_forbidden_lexical_names.contains(name))
+                    return;
+                if (name == identifier_group_name)
+                    scope_has_declaration = true;
+            }));
+            MUST(m_node->for_each_lexically_declared_name([&](auto const& name) {
+                if (name == identifier_group_name)
+                    scope_has_declaration = true;
+            }));
+
+            bool function_declaration = false;
+            MUST(m_node->for_each_var_function_declaration_in_reverse_order([&](auto const& declaration) {
+                if (declaration.name() == identifier_group_name)
+                    function_declaration = true;
+            }));
+            MUST(m_node->for_each_lexical_function_declaration_in_reverse_order([&](auto const& declaration) {
+                if (declaration.name() == identifier_group_name)
+                    function_declaration = true;
+            }));
+            MUST(m_node->for_each_function_hoistable_with_annexB_extension([&](auto const& declaration) {
+                if (declaration.name() == identifier_group_name)
+                    function_declaration = true;
+            }));
```

Comment:
It would be nice to break out of these loops immediately when `function_declaration` becomes `true`.

Reply:
that's true, but I would prefer to make this helpers to iterate declarations interruptible in a separate change: it looks like there are many places were these are used. I bet we will have to fix that anyway at some point if it will show up in profiles :)